In [8]:
import pandas as pd

df = pd.read_csv('../datasets/reviews.csv')

In [9]:
df.head()

id       asins   brand                  categories  \
0  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
1  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
2  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
3  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
4  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   

  colors             dateAdded           dateUpdated  \
0    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
1    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
2    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
3    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
4    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   

                  dimension  ean                         keys  ...  \
0  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
1  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
2  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
3  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
4  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   

  reviews.rating                                 reviews.sourceURLs  \
0            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
1            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
2            4.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
3            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
4            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   

                                        reviews.text  \
0  I initially had trouble deciding between the p...   
1  Allow me to preface this with a little history...   
2  I am enjoying it so far. Great for reading. Ha...   
3  I bought one of the first Paperwhites and have...   
4  I have to say upfront - I don't like coroporat...   

                                reviews.title reviews.userCity  \
0              Paperwhite voyage, no regrets!              NaN   
1           One Simply Could Not Ask For More              NaN   
2  Great for those that just want an e-reader              NaN   
3                    Love / Hate relationship              NaN   
4                                   I LOVE IT              NaN   

  reviews.userProvince    reviews.username  sizes upc     weight  
0                  NaN          Cristina M    NaN NaN  205 grams  
1                  NaN               Ricky    NaN NaN  205 grams  
2                  NaN       Tedd Gardiner    NaN NaN  205 grams  
3                  NaN              Dougal    NaN NaN  205 grams  
4                  NaN  Miljan David Tanic    NaN NaN  205 grams  

[5 rows x 27 columns]

In [10]:
print('Number of reviews: ', len(df))

Number of reviews:  1597


`reviews.csv` counts with 1597 review samples. Some of the columns are not relevant for our model trainning, relevant columns are:
- `reviews.rating`
- `reviews.title`
- `reviews.text`

Other columns can be dropped.

In [11]:
df = df.filter(['reviews.rating', 'reviews.title', 'reviews.text'], axis=1)
df.head(10)

reviews.rating                                      reviews.title  \
0             5.0                     Paperwhite voyage, no regrets!   
1             5.0                  One Simply Could Not Ask For More   
2             4.0         Great for those that just want an e-reader   
3             5.0                           Love / Hate relationship   
4             5.0                                          I LOVE IT   
5             NaN  Great device for reading. 8 people found this ...   
6             NaN  One Simply Could Not Ask For More 28 people fo...   
7             NaN  Definitely better than the previous generation...   
8             NaN  Paperwhite voyage, no regrets! 16 people found...   
9             NaN  Great for those that just want an e-reader 19 ...   

                                        reviews.text  
0  I initially had trouble deciding between the p...  
1  Allow me to preface this with a little history...  
2  I am enjoying it so far. Great for reading. Ha...  
3  I bought one of the first Paperwhites and have...  
4  I have to say upfront - I don't like coroporat...  
5  My previous kindle was a DX, this is my second...  
6  Allow me to preface this with a little history...  
7  Just got mine right now. Looks the same as the...  
8  I initially had trouble deciding between the p...  
9  I am enjoying it so far. Great for reading. Ha...

In [12]:
# rename collums to remove dots
df = df.rename({ 'reviews.rating': 'rating', 'reviews.title': 'title', 'reviews.text': 'text' }, axis=1)

In [13]:
df.head(10)

rating                                              title  \
0     5.0                     Paperwhite voyage, no regrets!   
1     5.0                  One Simply Could Not Ask For More   
2     4.0         Great for those that just want an e-reader   
3     5.0                           Love / Hate relationship   
4     5.0                                          I LOVE IT   
5     NaN  Great device for reading. 8 people found this ...   
6     NaN  One Simply Could Not Ask For More 28 people fo...   
7     NaN  Definitely better than the previous generation...   
8     NaN  Paperwhite voyage, no regrets! 16 people found...   
9     NaN  Great for those that just want an e-reader 19 ...   

                                                text  
0  I initially had trouble deciding between the p...  
1  Allow me to preface this with a little history...  
2  I am enjoying it so far. Great for reading. Ha...  
3  I bought one of the first Paperwhites and have...  
4  I have to say upfront - I don't like coroporat...  
5  My previous kindle was a DX, this is my second...  
6  Allow me to preface this with a little history...  
7  Just got mine right now. Looks the same as the...  
8  I initially had trouble deciding between the p...  
9  I am enjoying it so far. Great for reading. Ha...

In [14]:
# NaN ratings are not useful, so these should be removed
df.dropna(subset=['rating'])

rating                                              title  \
0        5.0                     Paperwhite voyage, no regrets!   
1        5.0                  One Simply Could Not Ask For More   
2        4.0         Great for those that just want an e-reader   
3        5.0                           Love / Hate relationship   
4        5.0                                          I LOVE IT   
...      ...                                                ...   
1592     3.0  I would be disappointed with myself if i produ...   
1593     1.0                        Battery draining remote!!!!   
1594     1.0      replacing an even worse remote. Waste of time   
1595     3.0                                         Overpriced   
1596     1.0  I am sending all of this crap back to amazon a...   

                                                   text  
0     I initially had trouble deciding between the p...  
1     Allow me to preface this with a little history...  
2     I am enjoying it so far. Great for reading. Ha...  
3     I bought one of the first Paperwhites and have...  
4     I have to say upfront - I don't like coroporat...  
...                                                 ...  
1592  This is not the same remote that I got for my ...  
1593  I have had to change the batteries in this rem...  
1594  Remote did not activate, nor did it connect to...  
1595  It does the job but is super over priced. I fe...  
1596  I ordered this item to replace the one that no...  

[1177 rows x 3 columns]

Our model should only classify english comments so predictions are more accurate. Reviews in other languages should be filtered out.

In [21]:
from langdetect import detect

def detect_english(title, text):
    if detect(title) == 'en' and detect(text) == 'en':
        return 'en'
    return 'unknown'

df['lang'] = df.apply(lambda x: detect_english(x.title, x.text), axis=1)

TypeError: expected string or bytes-like object

In [22]:
# Suspect theres some NaNs affecting `detect`
df.isna().sum()

rating    420
title      17
text        0
dtype: int64

In [23]:
df = df.dropna(subset=['rating'])
df.head(10)

rating                                              title  \
0      5.0                     Paperwhite voyage, no regrets!   
1      5.0                  One Simply Could Not Ask For More   
2      4.0         Great for those that just want an e-reader   
3      5.0                           Love / Hate relationship   
4      5.0                                          I LOVE IT   
13     4.0                             Liked the smaller size   
14     5.0  Superb reading device - but which one's best f...   
15     5.0                                         I love it!   
16     4.0                                         Un plaisir   
17     5.0          Works great and I love the built-in light   

                                                 text  
0   I initially had trouble deciding between the p...  
1   Allow me to preface this with a little history...  
2   I am enjoying it so far. Great for reading. Ha...  
3   I bought one of the first Paperwhites and have...  
4   I have to say upfront - I don't like coroporat...  
13  Had older model, that you could text to speech...  
14  This is a review of the Kindle Paperwhite laun...  
15  I love my kindle! I got one for my fiance on h...  
16  Vraiment bon petit appareil , lger et facile d...  
17  Exactly what it is supposed to be. Works great...

In [24]:
df = df.dropna(subset=['title'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1177 entries, 0 to 1596
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   rating  1177 non-null   float64
 1   title   1177 non-null   object 
 2   text    1177 non-null   object 
dtypes: float64(1), object(2)
memory usage: 36.8+ KB


In [25]:
df['lang'] = df.apply(lambda x: detect_english(x.title, x.text), axis=1)
df.head(10)

rating                                              title  \
0      5.0                     Paperwhite voyage, no regrets!   
1      5.0                  One Simply Could Not Ask For More   
2      4.0         Great for those that just want an e-reader   
3      5.0                           Love / Hate relationship   
4      5.0                                          I LOVE IT   
13     4.0                             Liked the smaller size   
14     5.0  Superb reading device - but which one's best f...   
15     5.0                                         I love it!   
16     4.0                                         Un plaisir   
17     5.0          Works great and I love the built-in light   

                                                 text     lang  
0   I initially had trouble deciding between the p...  unknown  
1   Allow me to preface this with a little history...       en  
2   I am enjoying it so far. Great for reading. Ha...       en  
3   I bought one of the first Paperwhites and have...       en  
4   I have to say upfront - I don't like coroporat...  unknown  
13  Had older model, that you could text to speech...       en  
14  This is a review of the Kindle Paperwhite laun...       en  
15  I love my kindle! I got one for my fiance on h...  unknown  
16  Vraiment bon petit appareil , lger et facile d...  unknown  
17  Exactly what it is supposed to be. Works great...       en

Seems like some of the rows now has `unkown`, this should be due to a limitation in `langdetect` or due to the exclusivity of the `detect_english`.

In [26]:
df['lang'].value_counts()

lang
en         950
unknown    227
Name: count, dtype: int64

Index 16 above is crearly French, but theres other cases like Index 15 where is English but there is a small percentage of French by using `fiance`, this might affect detection.

In [27]:
from langdetect import detect_langs

detect_langs('I love my kindle! I got one for my fiance on...')

[en:0.9999944771444464]

In [28]:
detect_langs('Vraiment bon petit appareil , lger et facile d...')

[fr:0.9999972864118577]

In [31]:
print(df.loc[[15]])

    rating       title                                               text  \
15     5.0  I love it!  I love my kindle! I got one for my fiance on h...   

       lang  
15  unknown  


Lets ease a bit the detection to allow either title or text to be English in order to determine wether the language is English, this could affect precision but might also bring in more samples to train the classification model.

In [32]:
from langdetect import detect

def detect_english(title, text):
    if detect(title) == 'en' or detect(text) == 'en':
        return 'en'
    return 'unknown'

df['lang'] = df.apply(lambda x: detect_english(x.title, x.text), axis=1)

In [33]:
df['lang'].value_counts()

lang
en         1173
unknown       4
Name: count, dtype: int64

This change brough in almost 98% of our samples! Now other items which are not English could be removed.

In [ ]:
df_filtered = df.query('a > 0').query('0 < b < 2')